rllib/algorithms/impala/impala_learner.py

```diff
@@ -175,15 +168,25 @@ def update_from_episodes(
                     self._gpu_loader_in_queue.qsize(),
                 )
             else:
-                # Enqueue to Learner thread's in-queue.
-                _LearnerThread.enqueue(
-                    self._learner_thread_in_queue,
-                    MultiAgentBatch(
-                        {mid: SampleBatch(b) for mid, b in batch.items()},
-                        env_steps=env_steps,
-                    ),
-                    self.metrics,
+                ma_batch = MultiAgentBatch(
+                    {mid: SampleBatch(b) for mid, b in batch.items()},
+                    env_steps=env_steps,
                 )
+                # Add the batch directly to the circular buffer.
+                if isinstance(self._learner_thread_in_queue, CircularBuffer):
```

Comment:
If the batch has already passed the learner connector, could there be any transformations in the connector that were made with a module that could become stale after some time of the batch in the buffer?

Reply:
Could be. I think that's by design, though (and v-trace is there for the defense :) ).
For example the paper talks about doing the target net forward pass on the batch already before adding it to the buffer. However, we don't do this, we perform a new target forward pass each time we compute a loss. I don't think it's crucial, though, that we do it this way. We would save a few (cheap, non-gradient) forward passes.